File: scripts/index_kb.ts
// Dev harness (Phase 7) — embed every bundled KB solution and upsert to Pinecone.
// Auto-creates the index if missing. Re-runnable (upserts overwrite by id).
// Run: node --env-file=.env.local --import tsx scripts/index_kb.ts
import { readdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { embed } from "../lib/llm";
import { hasPinecone, PINECONE_INDEX } from "../lib/env";
import { ensureIndex, kbIndex, type SolutionMetadata } from "../lib/kb/pinecone";

type Entry = {
  id: string; points: number; topic?: string; problem: string;
  official_solution: string; final_answer?: string; notes?: string;
};
type ExamKB = { exam: string; course: string; questions: Record<string, Entry> };

// What we embed: the QUESTION statement (topic + problem), so a student's
// transcribed work matches the problem it answers — not the solution text.
function embedText(e: Entry): string {
  return [e.topic, e.problem].filter(Boolean).join("\n").slice(0, 4000);
}

async function main() {
  if (!hasPinecone) {
    console.error("PINECONE_API_KEY not set — nothing to index.");
    process.exit(1);
  }
  const dir = join(process.cwd(), "lib", "kb", "solutions");
  const files = readdirSync(dir).filter((f) => f.endsWith(".json"));
  console.log(`KB files: ${files.join(", ") || "(none)"}`);

  // Stamp this run. Every solution vector written now carries it, and the Retriever
  // only trusts vectors from the generation recorded in generation.json.
  const gen = new Date().toISOString();

  const ids: string[] = [];
  const texts: string[] = [];
  const metas: SolutionMetadata[] = [];

  for (const f of files) {
    const kb = JSON.parse(readFileSync(join(dir, f), "utf-8")) as ExamKB;
    const slug = f.replace(/\.json$/, "");
    for (const e of Object.values(kb.questions)) {
      ids.push(`sol:${slug}:${e.id}`);
      texts.push(embedText(e));
      metas.push({
        kind: "solution",
        gen,
        entryId: e.id,
        points: e.points,
        topic: e.topic ?? "",
        problem: e.problem,
        official_solution: e.official_solution,
        final_answer: e.final_answer ?? "",
        notes: e.notes ?? "",
        exam: kb.exam,
        course: kb.course,
      });
    }
  }

  if (ids.length === 0) {
    console.error("No KB entries found.");
    process.exit(1);
  }
  console.log(`Embedding ${ids.length} entr${ids.length === 1 ? "y" : "ies"}…`);
  const vectors = await embed(texts);

  console.log(`Ensuring index "${PINECONE_INDEX}"…`);
  await ensureIndex();

  const records = ids.map((id, i) => ({ id, values: vectors[i], metadata: metas[i] }));
  await kbIndex().upsert({ records }); // v8: upsert takes { records }, not a bare array
  console.log(`Upserted ${records.length} vectors:`);
  records.forEach((r) => console.log(`  - ${r.id}  (${r.metadata.exam} · ${r.metadata.points} pts)`));

  // Record the generation the app should trust. Upsert overwrites by id but never
  // removes, and this index rejects delete-by-id, so a re-ingest that renames questions
  // leaves the old vectors in place. Writing the stamp here is what retires them.
  const keep = new Set(ids);
  let superseded = 0;
  let paginationToken: string | undefined;
  do {
    const page = await kbIndex().listPaginated({ prefix: "sol:", paginationToken });
    for (const v of page.vectors ?? []) if (v.id && !keep.has(v.id)) superseded++;
    paginationToken = page.pagination?.next;
  } while (paginationToken);

  writeFileSync(
    join(process.cwd(), "lib", "kb", "generation.json"),
    JSON.stringify({ gen, indexed: ids.length }, null, 2),
    { encoding: "utf-8" }
  );
  console.log(`Generation ${gen} — ${ids.length} live vector(s).`);
  if (superseded) {
    console.log(`${superseded} superseded vector(s) remain in the index but are now filtered out of retrieval.`);
  }

  // Give the serverless index a moment, then report the count.
  await new Promise((r) => setTimeout(r, 3000));
  const stats = await kbIndex().describeIndexStats();
  console.log("index total vectors:", stats.totalRecordCount);
}

main().catch((e) => {
  console.error("INDEX FAILED:", e);
  process.exit(1);
});

File: scripts/index_kb.py
"""Dev harness -- embed every bundled KB solution and upsert to Pinecone.
Port of scripts/index_kb.ts. Auto-creates the index if missing. Re-runnable.

Run: python scripts/index_kb.py
"""
from __future__ import annotations

import json
import os
import sys
import time
from datetime import datetime, timezone

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from checkmate.env import HAS_PINECONE, PINECONE_INDEX  # noqa: E402
from checkmate.kb.pinecone import ensure_index, kb_index  # noqa: E402
from checkmate.llm import embed  # noqa: E402

_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
_KB = os.path.join(_ROOT, "checkmate", "kb")
_SOL_DIR = os.path.join(_KB, "solutions")


def embed_text(e: dict) -> str:
    """What we embed: the QUESTION statement (topic + problem), so a student's transcribed
    work matches the problem it answers -- not the solution text."""
    return "\n".join(x for x in [e.get("topic"), e.get("problem")] if x)[:4000]


def main() -> None:
    if not HAS_PINECONE:
        print("PINECONE_API_KEY not set — nothing to index.")
        sys.exit(1)

    files = sorted(f for f in os.listdir(_SOL_DIR) if f.endswith(".json"))
    print(f"KB files: {', '.join(files) or '(none)'}")

    # Stamp this run. Every vector written now carries it; the Retriever only trusts vectors
    # from the generation recorded in generation.json.
    gen = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")

    ids: list[str] = []
    texts: list[str] = []
    metas: list[dict] = []

    for f in files:
        kb = json.load(open(os.path.join(_SOL_DIR, f), encoding="utf-8"))
        slug = f[:-5]
        for e in kb["questions"].values():
            ids.append(f"sol:{slug}:{e['id']}")
            texts.append(embed_text(e))
            metas.append({
                "kind": "solution", "gen": gen, "entryId": e["id"], "points": e.get("points", 0),
                "topic": e.get("topic", "") or "", "problem": e.get("problem", ""),
                "official_solution": e.get("official_solution", ""),
                "final_answer": e.get("final_answer", "") or "", "notes": e.get("notes", "") or "",
                "exam": kb["exam"], "course": kb["course"],
            })

    if not ids:
        print("No KB entries found.")
        sys.exit(1)

    print(f"Embedding {len(ids)} entr{'y' if len(ids) == 1 else 'ies'}…")
    vectors = embed(texts)

    print(f'Ensuring index "{PINECONE_INDEX}"…')
    ensure_index()

    index = kb_index()
    records = [{"id": ids[i], "values": vectors[i], "metadata": metas[i]} for i in range(len(ids))]
    index.upsert(vectors=records)
    print(f"Upserted {len(records)} vectors:")
    for r in records:
        print(f"  - {r['id']}  ({r['metadata']['exam']} · {r['metadata']['points']} pts)")

    # Record the generation the app should trust. Upsert overwrites by id but never removes,
    # so writing the stamp here is what retires vectors an earlier run left behind.
    keep = set(ids)
    superseded = 0
    token = None
    while True:
        page = index.list_paginated(prefix="sol:", pagination_token=token)
        page_vectors = getattr(page, "vectors", None) or (page.get("vectors") if isinstance(page, dict) else [])
        for v in page_vectors or []:
            vid = getattr(v, "id", None) or (v.get("id") if isinstance(v, dict) else None)
            if vid and vid not in keep:
                superseded += 1
        pagination = getattr(page, "pagination", None) or (page.get("pagination") if isinstance(page, dict) else None)
        token = getattr(pagination, "next", None) or (pagination.get("next") if isinstance(pagination, dict) else None)
        if not token:
            break

    with open(os.path.join(_KB, "generation.json"), "w", encoding="utf-8") as f:
        json.dump({"gen": gen, "indexed": len(ids)}, f, ensure_ascii=False, indent=2)
    print(f"Generation {gen} — {len(ids)} live vector(s).")
    if superseded:
        print(f"{superseded} superseded vector(s) remain in the index but are now filtered out of retrieval.")

    time.sleep(3)  # give the serverless index a moment
    stats = index.describe_index_stats()
    total = getattr(stats, "total_vector_count", None)
    if total is None and isinstance(stats, dict):
        total = stats.get("total_vector_count")
    print("index total vectors:", total)


if __name__ == "__main__":
    main()
